### PR TITLE
Chunked Prefill

### PR DESCRIPTION
**Pull request by hlahkar**

### CODE REVIEW

File: tests/basic_correctness/test_chunked_prefill.py
Comment (madamczykhabana) [inline tests/basic_correctness/test_chunked_prefill.py:1]: I don't think we should modify the test cases or at least add special case support for HPU.

@kzawora-intel what's the guideline around modifying test cases? Is this something we can do, we should avoid or any other option?

File: vllm/attention/backends/hpu_attn.py [hunk 1]
```diff
@@ -95,6 +98,12 @@ class HPUAttentionMetadata(HPUPagedAttentionMetadata, AttentionMetadata):
     cross_block_scales: Optional[torch.Tensor] = None
     cross_block_usage: Optional[torch.Tensor] = None
     cross_attn_bias: Optional[torch.Tensor] = None
+    decode_slot_mapping: Optional[torch.Tensor] = None
+    decode_block_list: Optional[torch.Tensor] = None
+    decode_block_indices: Optional[torch.Tensor] = None
+    decode_block_offsets: Optional[torch.Tensor] = None
+    decode_attn_bias: Optional[torch.Tensor] = None
+    chunk_prefill_enabled: bool = False
```
Comment (madamczykhabana) [inline vllm/attention/backends/hpu_attn.py:106]: The number of tensors in that dataclass is getting out of hand :(

File: vllm/attention/backends/hpu_attn.py [hunk 2]
```diff
@@ -163,6 +172,249 @@ def __init__(
...
+                                            device=device,
+                                            dtype=torch.bool),
+                                diagonal=1)
+        mask = causal_mask.logical_or(len_mask)
+        mask = torch.concat((past_mask, mask), dim=-1)
+        attn_bias = (torch.zeros_like(mask, dtype=dtype).masked_fill_(
+            mask, -math.inf))
+        return attn_bias
+
+    def forward_chunked_prefill(
```
Comment (madamczykhabana) [inline vllm/attention/backends/hpu_attn.py:199]: This function is huge and really hard to follow. Please split it up into smaller parts. Perhaps some of the logic can be reused from existing implementations.

File: vllm/attention/backends/hpu_attn.py [hunk 3]
```diff
@@ -163,6 +172,249 @@ def __init__(
...
+                seq_lens_t = attn_metadata.seq_lens_tensor
+                context_lens_t = attn_metadata.context_lens_tensor
+                query_lens_t = seq_lens_t - context_lens_t
+                attn_bias = None
+                for i in range(batch_size):
+                    single_attn_bias = self._set_attn_bias(int(seq_len), context_lens_t[i], query_lens_t[i], query.device, query.dtype)
+                    if attn_bias == None:
+                        attn_bias = single_attn_bias
+                    else:
+                        attn_bias = torch.cat((single_attn_bias, attn_bias), dim=0)
```
Comment (madamczykhabana) [inline vllm/attention/backends/hpu_attn.py:364]: This might be slow both in runtime and compilation times. For example there are scenarios where we use batch_size > 640 . Additionally recalculating attn_bias for each layer can take a significant portion of time.

File: vllm/engine/arg_utils.py [hunk 4]
```diff
@@ -119,7 +119,7 @@ class EngineArgs:
     enable_prefix_caching: Optional[bool] = None
     disable_sliding_window: bool = False
     use_v2_block_manager: bool = True
-    use_padding_aware_scheduling: bool = current_platform.is_hpu()
+    use_padding_aware_scheduling: bool = True
```
Comment (madamczykhabana) [inline vllm/engine/arg_utils.py:122]: We shouldn't change the default for other platforms

File: vllm/model_executor/sampling_metadata.py [hunk 5]
```diff
@@ -291,8 +291,7 @@ def _prepare_seq_groups(
         else:
             # Decode
             prompt_logprob_len = 0
-            query_len = query_lens[i] if query_lens is not None and len(
-                query_lens) > 0 else 1
+            query_len = 1
```
Comment (madamczykhabana) [inline vllm/model_executor/sampling_metadata.py:294]: Is it ok to always use 1?

File: vllm/worker/hpu_model_runner.py
Comment (madamczykhabana) [inline vllm/worker/hpu_model_runner.py:1]: Adding some much duplicated code will be a huge maintenance burden. We need to figure out a better way to structure the code and reuse as much as possible

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

File: vllm/attention/backends/hpu_attn.py [hunk 6]
```diff
@@ -163,6 +172,249 @@ def __init__(
...
+                    prefill_value.view(kv_shape),
+                    attn_bias=attn_bias,
+                    p=0.0,
+                    scale=self.scale,
+                    matmul_qk_op=self.matmul_qk,
+                    softmax_op=self.softmax,
+                    matmul_av_op=self.matmul_av,
+                    valid_seq_lengths=attn_metadata.seq_lens_tensor,
+                    fsdpa_op=self.fused_scaled_dot_product_attention,
+                )
```
Comment (kzawora-intel) [inline vllm/attention/backends/hpu_attn.py:352]: In chunked prefill it is possible that a request gets split across engine iterations, e.g. for 1024 tokens, you might need to compute the first 512 in the first iteration, and the remaining 512 in second with prefix prefill. Does this handle that scenario properly?

**Comment by darekkreft** [comment]
@hlahkar I tested throughput_llama3_1_8b_128_128-bf16-1x-perf-vllm-g2-chunked-prefill with commit e29c70e6af9d9432f4126772791a881cdf7f35fe. Look this functionality not works functional yet.
From server got:

 IndexError: tuple index out of range, log ->[log]( https://qa-jenkins-ctrl03.habana-labs.com/job/qa_jobs/job/qa_testers/job/gdn-qa/job/pytorch/job/gaudi2/job/archived/job/VLLM/job/continous_batching/job/NVIDIA_like_benchmarks/job/chunked_prefill/job/on_demand_throughput_llama3_1_8b_128_128-bf16-1x-perf-vllm-g2-chunked-prefill_pytorch_gaudi2-ak8s_master/9/artifact/vllm_server.log)

From pytorch_log.txt ->[log](https://qa-jenkins-ctrl03.habana-labs.com/job/qa_jobs/job/qa_testers/job/gdn-qa/job/pytorch/job/gaudi2/job/archived/job/VLLM/job/continous_batching/job/NVIDIA_like_benchmarks/job/chunked_prefill/job/on_demand_throughput_llama3_1_8b_128_128-bf16-1x-perf-vllm-g2-chunked-prefill_pytorch_gaudi2-ak8s_master/9/artifact/habana_logs/pytorch_log.txt)

td::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*)+0x2d) [0x7f118f8ade1d]
[19:39:46.404082][PT_BRIDGE       ][error][tid:1BCE] /lib/x86_64-linux-gnu/libc.so.6(+0x99ee8) [0x7f1232538ee8]

File: tests/basic_correctness/test_chunked_prefill.py
Comment (hlahkar) [reply tests/basic_correctness/test_chunked_prefill.py:1]: @madamczykhabana I have added a new file for our test case; keeping the original test case as it is.

File: vllm/attention/backends/hpu_attn.py (hunk 2 above)
Comment (hlahkar) [reply vllm/attention/backends/hpu_attn.py:243]: @madamczykhabana This function I have  refactored into 2; one for the preprocessing and the second one for the actual attention. Please check once

File: vllm/attention/backends/hpu_attn.py (hunk 3 above)
Comment (hlahkar) [reply vllm/attention/backends/hpu_attn.py]: set_attn_bias has been moed to hpu_model_runner; and is now calculated only once per batch

File: vllm/model_executor/sampling_metadata.py (hunk 5 above)
Comment (hlahkar) [reply vllm/model_executor/sampling_metadata.py]: This code went in by mistake; and has been reverted.

File: vllm/worker/hpu_model_runner.py
Comment (hlahkar) [reply vllm/worker/hpu_model_runner.py:1]: Most of the duplicate code has been removed. Kindl check

File: vllm/attention/backends/hpu_attn.py (hunk 6 above)
Comment (hlahkar) [reply vllm/attention/backends/hpu_attn.py]: Since we are passing the attention_bias; it is taken care of

File: vllm/attention/backends/hpu_attn.py (hunk 1 above)
Comment (hlahkar) [reply vllm/attention/backends/hpu_attn.py:106]: I understand this; but unfortunately we need this information for mixed batches

File: tests/basic_correctness/test_chunked_prefill_hpu.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+"""Compare the outputs of HF and vLLM when using greedy sampling.
```
Comment (michalkuligowski) [inline tests/basic_correctness/test_chunked_prefill_hpu.py:1]: Please fix precommit issues.
Are those tests that @darekkreft  mentioned working now?

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: tests/basic_correctness/test_chunked_prefill_hpu.py (hunk 7 above)
Comment (hlahkar) [reply tests/basic_correctness/test_chunked_prefill_hpu.py:2]: All the precommit issues are solved and UT cases are passing. We are working on the issue mentioned by @darekkreft

**Comment by hlahkar** [comment]
As we are not taking this version of chunked prefill, closing this
